latch_cli/nextflow/workflow.py

```diff
@@ -0,0 +1,293 @@
+from latch.types.file import LatchFile
+from latch.types.directory import LatchDir, LatchOutputDir
+from latch.ldata.path import LPath
+from latch_cli.nextflow.workflow import get_flag
+from latch_cli.nextflow.utils import _get_execution_name
+from latch_cli.utils import urljoins
+from latch.types import metadata
+from flytekit.core.annotation import FlyteAnnotation
+
+import latch_metadata
+
+@custom_task(cpu=0.25, memory=0.5, storage_gib=1)
+def initialize() -> str:
+    token = os.environ.get("FLYTE_INTERNAL_EXECUTION_ID")
+    if token is None:
+        raise RuntimeError("failed to get execution token")
+
+    headers = {{"Authorization": f"Latch-Execution-Token {{token}}"}}
+
+    print("Provisioning shared storage volume... ", end="")
+    resp = requests.post(
+        "http://nf-dispatcher-service.flyte.svc.cluster.local/provision-storage",
+        headers=headers,
+        json={{
+            "storage_gib": {storage_gib},
+        }}
+    )
+    resp.raise_for_status()
+    print("Done.")
+
+    return resp.json()["name"]
+
+
+{preambles}
+
+{samplesheet_funs}
+
+@nextflow_runtime_task(cpu={cpu}, memory={memory})
+def nextflow_runtime(pvc_name: str, {param_signature}) -> None:
+    try:
+        shared_dir = Path("/nf-workdir")
+
+{samplesheet_constructors}
+
+        shutil.copytree(
+            Path("/root"),
+            shared_dir,
+            ignore=lambda src, names: ["latch", ".latch", "nextflow", ".nextflow", "work", "results"],
+            ignore_dangling_symlinks=True,
+            dirs_exist_ok=True,
+        )
+
+        cmd = [
+            "/root/nextflow",
+            "run",
+            str(shared_dir / "{nf_script}"),
+            "-work-dir",
+            str(shared_dir),
+            "-profile",
+            "{execution_profile}",
+            "-c",
+            "latch.config",
+{params_to_flags}
+        ]
+
+        print("Launching Nextflow Runtime")
+        print(' '.join(cmd))
+        print(flush=True)
+
+        env = {{
+            **os.environ,
+            "NXF_HOME": "/root/.nextflow",
+            "K8_STORAGE_CLAIM_NAME": pvc_name,
+        }}
+        subprocess.run(
+            cmd,
+            env=env,
+            check=True,
+            cwd=str(shared_dir),
+        )
+    except subprocess.CalledProcessError:
+        remote = LPath(urljoins("{remote_output_dir}", _get_execution_name(), "nextflow.log"))
+        print()
+        print(f"Uploading .nextflow.log to {{remote.path}}")
+        remote.upload_from(shared_dir / ".nextflow.log")
+        import time
+        time.sleep(10 * 60)
+        raise
+    finally:
+        token = os.environ.get("FLYTE_INTERNAL_EXECUTION_ID")
+        if token is None:
+            raise RuntimeError("failed to get execution token")
+
+        headers = {{"Authorization": f"Latch-Execution-Token {{token}}"}}
+        resp = requests.post(
+            "http://nf-dispatcher-service.flyte.svc.cluster.local/finalize",
```

Comment:
should also have a quota system / and or some sort of rate limiting on the dispatcher so that if someone decides to delete this they can't add more pods beyond a certain point

Reply:
will add to backlog. 
The finalize call likely won't be necessary once we write the async process that listens for updates and cleans up pods/pvc on workflow completion